Share convert 2

bendecoste opened the pull request.

bendecoste commented:
Requires CHANGELOG update

mortendahl commented on tf_encrypted/tensor/prime.py:
```diff
@@ -125,6 +129,9 @@ def _lift(x, y) -> Tuple[PrimeTensor, PrimeTensor]:
     if isinstance(x, int) and isinstance(y, PrimeTensor):
         return y.factory.tensor(np.array([x])), y
 
+    # if isinstance(x, PrimeTensor) and isinstance(y, OddImplicitTensor):
```
left overs?

mortendahl commented on tf_encrypted/tensor/prime.py:
```diff
@@ -7,6 +7,7 @@
 
 from .factory import AbstractFactory, AbstractTensor, AbstractConstant, AbstractVariable, AbstractPlaceholder
 from .shared import binarize, im2col
+# from .odd_implicit import OddImplicitTensor
```
left overs?

mortendahl commented on tf_encrypted/tensor/int64.py:
```diff
@@ -223,6 +224,29 @@ def squeeze(self, axis: Optional[List[int]]=None) -> 'Int64Tensor':
     def negative(self) -> 'Int64Tensor':
         return Int64Tensor(tf.negative(self.value))
 
+    def to_odd_modulus(self, factory: OddImplicitFactory) -> OddImplicitTensor:
```
let's use casting instead, ie along the lines of https://github.com/mortendahl/tf-encrypted/blob/master/tf_encrypted/tensor/prime.py#L112-L113

mortendahl commented on tf_encrypted/tensor/int32.py:
```diff
@@ -177,6 +178,12 @@ def equal(self, other, factory: AbstractFactory=int32factory) -> 'AbstractTensor
+    def compute_wrap(self, y: AbstractTensor, modulus: int) -> AbstractTensor:
+        return Int32Tensor(tf.cast(self.value + y.value >= modulus, dtype=tf.int32))
+
+    def to_odd_modulus(self, factory: OddImplicitFactory) -> OddImplicitTensor:
```
see comment on casting elsewhere

mortendahl commented on tf_encrypted/tensor/int32.py:
```diff
@@ -177,6 +178,12 @@ def equal(self, other, factory: AbstractFactory=int32factory) -> 'AbstractTensor
         return factory.tensor(tf.cast(tf.equal(x.value, y.value), dtype=factory.native_type))
 
+    def compute_wrap(self, y: AbstractTensor, modulus: int) -> AbstractTensor:
+        return Int32Tensor(tf.cast(self.value + y.value >= modulus, dtype=tf.int32))
```
won't this overflow if we're working with int32s?

mortendahl commented on tf_encrypted/protocol/securenn.py:
```diff
@@ -31,6 +32,7 @@ def __init__(
         server_2: Optional[Player] = None,
         prime_factory: Optional[AbstractFactory] = None,
         odd_factory: Optional[AbstractFactory] = None,
+        implicit_factory: Optional[AbstractFactory] = None,
```
I think `odd_factory` was intended for this!

mortendahl commented on tf_encrypted/protocol/securenn.py:
```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
+
+        # line 10, j = 0
+        with tf.device(self.server_0.device_name):
+            n_0 = np_odd.share0 + npp.value_on_0 - (np_odd.share0 * npp.value_on_0 * 2)
```
this looks like an xor

mortendahl commented on tf_encrypted/protocol/securenn.py:
```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
+        with tf.device(self.server_0.device_name):
+            n_0 = np_odd.share0 + npp.value_on_0 - (np_odd.share0 * npp.value_on_0 * 2)
+            a_0 = a.share0.to_odd_modulus(self.implicit_factory)
+            theta_0 = beta_0.to_odd_modulus(self.implicit_factory) + (((-alpha).to_odd_modulus(self.implicit_factory) - 1) * 1) + gamma.share0 + n_0
```
I think you should be able to compute `theta` without using `tf.device`

bendecoste replied:
I think we'd need to write casting for pond tensors which I can do but might be a little out of scope here.

mortendahl commented on tf_encrypted/protocol/securenn.py:
```diff
@@ -299,17 +306,104 @@ def equal_zero(self, x, out_dtype: Optional[AbstractFactory]=None):
+
+    def compute_wrap(self, s0, s1, L: int):
+        # classical overflow
+        overflow_max = tf.cast(tf.logical_and((s0.value > 0), (s1.value > tf.int64.max - s0.value)), dtype=tf.bool)
```
you should be able to get `int64` from the factory so that this is not hardcoded

mortendahl commented on tests/test_share_convert.py:
```diff
@@ -0,0 +1,45 @@
+            tensor_factory=int64factory
+        )
+
+        val_a = np.array([100, -100])
```
could we try with a bunch of random values instead? i'm worried this won't catch corner cases

bendecoste commented:
@mortendahl I addressed most PR comments or replied to them.

I actually uncovered what I think is a bug.  If you try to share convert a sufficiently large number it will fail ~5% of the time.  I changed the test to show this.  Not sure it's worth fixing right now because we are thinking that we can bypass share convert anyway.  Considering just deferring to a different issue?

mortendahl commented:
believe we can close this guy per https://github.com/mortendahl/tf-encrypted/pull/306

mortendahl commented:
@bendecoste where did we end up re failing 5% of the time?

mortendahl commented:
any thoughts @bendecoste?

mortendahl commented:
closing since we don't need share convert for current applications and errors was never solved
